### ISSUE TITLE
Expired partition tombstones are kept in cache indefinitely - 

### ISSUE THREAD

**Issue by mykaul** [scylladb/scylladb#16092]
Follow-up to https://github.com/scylladb/scylladb/issues/6033 - to deal specifically with partition tombstones.

**Comment by bhalevy** [comment]
@tgrabiec/@michoecho (and @xemul who changed the cache tree implementations), please shed more light on this issue.
What exactly needs to be fixed and what is the plan to fix it.

Also, are the partition tombstones subject to lru, and if so, what's the penality for keeping them around while the partition is frequently accessed?

**Comment by michoecho** [comment]
I'm not sure what the issue is supposed to be. 

> are the partition tombstones subject to lru

Partition tombstone aren't separate entries in cache. They are a member field of `partition_version`. They are subject to lru in the sense that they are evicted when the whole partition is evicted.

> what's the penality for keeping them around while the partition is frequently accessed

I don't see any meaningful penalty.

**Comment by xemul** [comment]
> Partition tombstone aren't separate entries in cache. They are a member field of partition_version. They are subject to lru in the sense that they are evicted when the whole partition is evicted.

Maybe the intent was to track _range_ tombstones? These are kept in cache and are not (were not) subject to LRU eviction unlike e.g. row entries

**Comment by michoecho** [comment]
> > Partition tombstone aren't separate entries in cache. They are a member field of partition_version. They are subject to lru in the sense that they are evicted when the whole partition is evicted.
> 
> Maybe the intent was to track _range_ tombstones? These are kept in cache and are not (were not) subject to LRU eviction unlike e.g. row entries

No, range tombstones are already dealt with. They are subject to LRU since `mutation_partition_v2`, and indefinite pollution should be prevented by https://github.com/scylladb/scylladb/pull/14463.

But, I've responded before actually reading https://github.com/scylladb/scylladb/issues/6033. (I thought 6033 was only about range tombstones, but apparently since the beginning it was about partition tombstones too). And IIUC the argument there is that we should remove *any* expired tombstones from the cache, including partition tombstones, not for the sake of performance, but for the sake of correctness. As in: the user has a right to expect that expiration is reliable, i.e. that an expired tombstone won't affect new writes, even if their timestamps are lower than the tombstone's timestamp.

I think that's just wrong. The entire database was written with the contrary assumption. In particular, there is no way to distinguish an old write from a new write with the same timestamp.

If you wanted to get a guarantee that all expired tombstones are purged, you would probably need some new command that compacts the entire dataset on the node (to ensure that the expired tombstones remain applied to all *old* writes they were supposed to remove) and then purges the expired tombstones.

**Comment by avikivity** [comment]
The tombstones-in-cache tombstone is when a row or partition is composed only of tombstones. Let's start with cell tombstones first. If a row is completely composed of tombstones and is part of a frequently-read partition, it will be kept in cache indefinitely. As other cells are deleted, the problem grows. Eventually we have a large partition in cache, even though the live part is small.

For partition tombstones, the problem is similar but affects partition scans. If the table mostly has dead partitions but is scanned frequently, then those dead partitions will be kept in cache indefinitely, slowing down reads.

**Comment by avikivity** [comment]
> > > Partition tombstone aren't separate entries in cache. They are a member field of partition_version. They are subject to lru in the sense that they are evicted when the whole partition is evicted.
> > 
> > 
> > Maybe the intent was to track _range_ tombstones? These are kept in cache and are not (were not) subject to LRU eviction unlike e.g. row entries
> 
> No, range tombstones are already dealt with. They are subject to LRU since `mutation_partition_v2`, and indefinite pollution should be prevented by #14463.
> 
> But, I've responded before actually reading #6033. (I thought 6033 was only about range tombstones, but apparently since the beginning it was about partition tombstones too). And IIUC the argument there is that we should remove _any_ expired tombstones from the cache, including partition tombstones, not for the sake of performance, but for the sake of correctness. As in: the user has a right to expect that expiration is reliable, i.e. that an expired tombstone won't affect new writes, even if their timestamps are lower than the tombstone's timestamp.
> 
> I think that's just wrong. The entire database was written with the contrary assumption. In particular, there is no way to distinguish an old write from a new write with the same timestamp.


Agree.

> 
> If you wanted to get a guarantee that all expired tombstones are purged, you would probably need some new command that compacts the entire dataset on the node (to ensure that the expired tombstones remain applied to all _old_ writes they were supposed to remove) and then purges the expired tombstones.

That's not a requirement. We want to get rid of expired tombstones in cache in order to not slow down reads of adjacent rows.
